### PR TITLE
Add a dense layer with bias implementation

### PR DESCRIPTION
**Pull request by hugoabonizio**

### CODE REVIEW

File: t5x/examples/decoder_only/layers.py [hunk 1]
```diff
@@ -557,6 +557,72 @@ def __call__(self, inputs: Array) -> Array:
     return lax.dot_general(inputs, kernel, ((axis, contract_ind), ((), ())))
 
 
+#------------------------------------------------------------------------------
+# DenseGeneral for attention layers.
+#------------------------------------------------------------------------------
+class DenseWithBias(nn.Module):
+  """A linear transformation (without bias) with flexible axes.
```
Comment (rodrigonogueira4) [inline t5x/examples/decoder_only/layers.py:564]: remove (without bias)?

File: t5x/examples/decoder_only/layers.py [hunk 2]
```diff
@@ -557,6 +557,72 @@ def __call__(self, inputs: Array) -> Array:
     return lax.dot_general(inputs, kernel, ((axis, contract_ind), ((), ())))
 
 
+#------------------------------------------------------------------------------
+# DenseGeneral for attention layers.
```
Comment (rodrigonogueira4) [inline t5x/examples/decoder_only/layers.py:561]: DenseWithBias for FFN layers?

**Comment by rodrigonogueira4** [comment]
Thanks for implementing this! Do you know why we cannot use the nn.Linear from FLAX Or why T5x folks didn't use it?

File: t5x/examples/decoder_only/layers.py [hunk 3]
```diff
@@ -557,6 +557,72 @@ def __call__(self, inputs: Array) -> Array:
...
+      kernel_init: initializer function for the weight matrix.
+      bias_init: initializer function for the bias.
+  """
+  features: Union[Iterable[int], int]
+  axis: Union[Iterable[int], int] = -1
+  dtype: DType = jnp.float32
+  kernel_init: Initializer = nn.initializers.variance_scaling(
+      1.0, 'fan_in', 'truncated_normal')
+  kernel_axes: Tuple[str, ...] = ()
+  bias_init: Initializer = nn.initializers.ones
```
Comment (rodrigonogueira4) [inline t5x/examples/decoder_only/layers.py:579]: Could you please check if bias should be zero?

Looking at HF implementation, it should be zeros:
https://github.com/huggingface/transformers/blob/94b3f544a1f5e04b78d87a2ae32a7ac252e22e31/src/transformers/models/gptj/modeling_flax_gptj.py#L306
https://flax.readthedocs.io/en/latest/_modules/flax/linen/linear.html#Dense

Also in the mesh-transformer-jax:

https://github.com/kingoflolz/mesh-transformer-jax/blob/0a75ca9370576ad9d247facf6cb8e9699300e690/mesh_transformer/layers.py#L256

https://dm-haiku.readthedocs.io/en/latest/api.html?highlight=linear#id1

File: t5x/examples/decoder_only/layers.py [hunk 4]
```diff
@@ -587,6 +653,7 @@ class MlpBlock(nn.Module):
   activations: Sequence[Union[str, Callable]] = ('relu',)
   kernel_init: Initializer = nn.initializers.variance_scaling(
       1.0, 'fan_in', 'truncated_normal')
+  bias_init: Initializer = nn.initializers.ones
```
Comment (rodrigonogueira4) [inline t5x/examples/decoder_only/layers.py:656]: Note to change here in case bias turns out to be zeros in the original GPT-J implementation.

File: t5x/examples/decoder_only/network.py [hunk 5]
```diff
@@ -218,10 +218,11 @@ def __call__(self,
       logits = logits / jnp.sqrt(y.shape[-1])
     else:
       # Use a separate dense layer for the logit transform.
-      logits = layers.DenseGeneral(
+      logits = layers.DenseWithBias(
           cfg.vocab_size,
           dtype=jnp.float32,  # Use float32 for stabiliity.
           kernel_axes=('embed', 'vocab'),
+          bias_axes=('embed',),
```
Comment (rodrigonogueira4) [inline t5x/examples/decoder_only/network.py:225]: should be in the vocab axis?

**Comment by hugoabonizio** [comment]
> Thanks for implementing this! Do you know why we cannot use the nn.Linear from FLAX Or why T5x folks didn't use it?

@rodrigonogueira4 the problem is the same as in LayerNorm, using the Flax implementation it complains about the axes names. I think the problem is the lack of `param_with_axes` in parameter creation, but I'm not 100% sure.

File: t5x/examples/decoder_only/layers.py (hunk 3 above)
Comment (hugoabonizio) [reply t5x/examples/decoder_only/layers.py:579]: You're right, I missed that!

**Comment by rodrigonogueira4** [comment]
Got it, thanks!

File: t5x/examples/decoder_only/layers.py [hunk 6]
```diff
@@ -557,6 +557,72 @@ def __call__(self, inputs: Array) -> Array:
...
+    kernel = param_with_axes(
+        'kernel',
+        self.kernel_init,
+        kernel_param_shape,
+        jnp.float32,
+        axes=self.kernel_axes)
+    kernel = jnp.asarray(kernel, self.dtype)
+    kernel = jnp.reshape(kernel, kernel_shape)
+
+
```
Comment (rodrigonogueira4) [inline t5x/examples/decoder_only/layers.py:610]: nit: remove extra line?

File: t5x/examples/decoder_only/network.py [hunk 7]
```diff
@@ -218,10 +218,11 @@ def __call__(self,
       logits = logits / jnp.sqrt(y.shape[-1])
     else:
       # Use a separate dense layer for the logit transform.
-      logits = layers.DenseGeneral(
+      logits = layers.DenseWithBias(
           cfg.vocab_size,
           dtype=jnp.float32,  # Use float32 for stabiliity.
           kernel_axes=('embed', 'vocab'),
+          bias_axes=('vocab',),
```
Comment (rodrigonogueira4) [inline t5x/examples/decoder_only/network.py:225]: maybe also pass `bias_init` here? I know it is the same as the default one, but just in case we change the default...

Comment (hugoabonizio) [reply t5x/examples/decoder_only/network.py:225]: In this case we'd also need to include a `kernel_init` for consistency, but since the original code doesn't define this maybe we can leave it as is?

**Review by piresramon — APPROVED** [review]

File: t5x/examples/decoder_only/network.py (hunk 7 above)
Comment (rodrigonogueira4) [reply t5x/examples/decoder_only/network.py:225]: Since we want to reproduce GPT-J's training exactly, maybe pass their kernel_init and bias_init to this function?

Comment (rodrigonogueira4) [reply t5x/examples/decoder_only/network.py:225]: Please disregard this comment

**Review by ZanezZephyrs — APPROVED** [review]
